key_gen.go

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tally
+
+import (
+	"bytes"
+	"sort"
+)
+
+const (
+	prefixSplitter  = '+'
+	keyPairSplitter = ','
+	keyNameSplitter = '='
+)
+
+var (
+	keyGenPool = newKeyGenerationPool(1024, 1024, 32)
+	nilString  = ""
+)
+
+type keyGenerationPool struct {
+	bufferPool  *ObjectPool
+	stringsPool *ObjectPool
+}
+
+// KeyForStringMap generates a unique key for a map string set combination.
+func KeyForStringMap(
+	stringMap map[string]string,
+) string {
+	keys := keyGenPool.stringsPool.Get().([]string)
+	for k := range stringMap {
+		keys = append(keys, k)
+	}
+
+	buf := keyGenPool.bufferPool.Get().(*bytes.Buffer)
+	writeKeyValuesSortedByKeys(buf, keys, stringMap)
+
+	key := buf.String()
+	keyGenPool.release(buf, keys)
+	return key
+}
+
+// KeyForPrefixedStringMap generates a unique key for a
+// a prefix and a map string set combination.
+func KeyForPrefixedStringMap(
```

Comment:
Can KeyForStringMap just call this with empty prefix and we do a check in this func?